### DISCUSSION TITLE
linking prompts with tracing  in langchain integration

### DISCUSSION THREAD

**Discussion by SatheeshJM** [langfuse/langfuse#1895] [Ideas]
### Describe the feature or potential improvement

The documentation under "Prompt Management" (https://langfuse.com/docs/prompts/get-started#link-with-langfuse-tracing-optional) says `This is currently unavailable when using the LangChain or LlamaIndex integration`.

<img width="844" alt="image" src="https://github.com/langfuse/langfuse/assets/1222562/969e9ed7-30c4-4985-866f-fc6c4afde318">

Any plans on supporting this in the near future?

The prompt manager feature is pretty cool, but seems incomplete without a langchain integration

**Comment by marcklingen** [comment]
Currently exploring options how to make this work as prompts are then used deeply nested within a Llamaindex/Langchain trace and are not easy to detect. Do you have any ideas?

Fyi, you can use prompt management without linking traces to prompts even though being able to link is really useful.

**Reply by ajram23** [reply]
Is this why the generation count is zero when using prompts in Langchain?

**Reply by marcklingen** [reply]
yes!

**Reply by aiakubovich** [reply]
I was looking at some similar question about outputting final prompt with Langchain. There is some discussion about it: https://github.com/langchain-ai/langchain/issues/912. Not sure if it helps or not with Langchain integration.

**Reply by marcklingen** [reply]
thanks for sharing!

**Comment by ajram23** [comment]
@marcklingen Curious, Will this be prioritized anytime soon?

**Reply by marcklingen** [reply]
Generally yes, but not timeline as currently a lot of effort goes into v3 infra (scaling), manual annotation workflow features, and improvements to evals. Will keep you in the loop here.

**Comment by hassiebp** [comment]
Linking Langfuse prompts to Langchain executions has been one of the most requested features recently, and we have just released support for it 🫡 🥳

For more details, please check out:

- [Docs on how to integrate, see tabs for Langchain in examples](https://langfuse.com/docs/prompts/get-started#link-with-langfuse-tracing-optional)
- [Python SDK release](https://github.com/langfuse/langfuse-python/releases/tag/v2.49.0)
- [JS SDK release](https://github.com/langfuse/langfuse-js/releases/tag/v3.25.0)

**Reply by ajram23** [reply]
@hassiebp very cool. I just tried it out and I am not sure if it's working. The generation counts havent changed, nor do I see it in the traces. I have link for you if you need to debug. Not sure if I am making a mistake. I am using Python SDK and I am on langchain 0.2.X

**Reply by hassiebp** [reply]
> @hassiebp very cool. I just tried it out and I am not sure if it's working. The generation counts havent changed, nor do I see it in the traces. I have link for you if you need to debug. Not sure if I am making a mistake.

Thanks for trying it out @ajram23! Could you please create an issue with a minimum repro and the Langfuse + Langchain versions you are using?

**Reply by ajram23** [reply]
@hassiebp Got other higher priority stuff that needs my attention now. Here is the exact list of versions langchain==0.2.1
langchain-anthropic==0.1.20
langchain-community==0.2.1
langchain-core==0.2.17
langchain-groq==0.1.6
langchain-openai==0.1.14
langchain-text-splitters==0.2.1
langfuse==2.49.0 I can also provide you a link to a trace within Langfuse so you can check if its getting the metadata or not. I can come back to this later in the week and provide a repo. Also what is not clear from the documentation is how do you handle ChatPromptTemplate.from_messages, FewShotPromptTemplate which pass in prompts as parameters. Thanks!

**Reply by hassiebp** [reply]
> @hassiebp Got other higher priority stuff that needs my attention now. Here is the exact list of versions langchain==0.2.1
> 
> langchain-anthropic==0.1.20
> 
> langchain-community==0.2.1
> 
> langchain-core==0.2.17
> 
> langchain-groq==0.1.6
> 
> langchain-openai==0.1.14
> 
> langchain-text-splitters==0.2.1
> 
> langfuse==2.49.0 I can also provide you a link to a trace within Langfuse so you can check if its getting the metadata or not. I can come back to this later in the week and provide a repo. Also what is not clear from the documentation is how do you handle ChatPromptTemplate.from_messages, FewShotPromptTemplate which pass in prompts as parameters. Thanks!

Thanks, @ajram23 - if you can send me the trace to hassieb [at] langfuse [dot] com I can take a closer look

**Reply by ajram23** [reply]
done! Sent!

**Reply by hassiebp** [reply]
Thanks @ajram23 ! 

As a rule here, the metadata of prompt object must hold a dict that has a `langfuse_prompt` key with the langfuse prompt client object as its value. In Langchain, the `metadata` property exists on all prompt templates inheriting from its `BasePromptTemplate` ([see here](https://api.python.langchain.com/en/latest/prompts/langchain_core.prompts.base.BasePromptTemplate.html#langchain_core.prompts.base.BasePromptTemplate.metadata))

So if the `.from_messages` method does not accept a metadata kwarg, you can always manually set it in an imperative call: 

```python /"langfuse_prompt"/
from langfuse import Langfuse
from langchain_core.prompts import ChatPromptTemplate, PromptTemplate
from langchain_openai import ChatOpenAI, OpenAI

langfuse = Langfuse()

# Chat prompts:
langfuse_chat_prompt = langfuse.get_prompt("movie-critic-chat", type="chat")

## Manually set the metadata on the langchain_chat_prompt to link it to generations that use it
langchain_chat_prompt = ChatPromptTemplate.from_messages(
    langfuse_chat_prompt.get_langchain_prompt()
)
langchain_chat_prompt.metadata = {"langfuse_prompt": langfuse_chat_prompt}

## Use the chat prompt in a Langchain chain
chat_llm = ChatOpenAI()
chat_chain = langchain_chat_prompt | chat_llm

chat_chain.invoke({"movie": "Dune 2", "criticlevel": "expert"})
```

I have a PR open to update our docs with that example.

**Reply by pwnywiz** [reply]
i am facing the same issue with @ajram23 and im using `ChatPromptTemplate.from_messages()` just like your latest example. The problem is, if there are multiple messages, is there a way to set them all in a single value with `langfuse_prompt` as key?

This is a current example code where prompts arent getting linked to the traces:
```
ChatPromptTemplate.from_messages(
    [
        SystemMessagePromptTemplate.from_template(
            my_prompt.prompt,
            metadata={"langfuse_prompt": my_prompt},
        ),
        HumanMessagePromptTemplate.from_template(another_prompt.prompt, metadata={"langfuse_prompt": another_prompt}),
    ]
)
```

Langfuse version: 2.49.0
Langchain version: 0.2.14

**Reply by hassiebp** [reply]
@pwnywiz The relationship between a generation in Langfuse and a prompt is one-to-one, so having a generation linked to two prompts (one for the system message, one for the user message) is not possible.

Also, the `langfuse_prompt` on the `metadata` must be on the prompt template that is a node in the final chain. Having them inside the `from_messages` will not work in the current callback implementation of Langchain.

So for your case, you can choose one prompt that you'd like to link and set it to the metadata of the resulting ChatPromptTemplate.

**Reply by marcklingen** [reply]
Shall we move the 1:n relationship to another thread? This could be a nice addition irrespective of the Langchain <> Prompt Management integration.

There is also this related thread about prompt composability which would reduce the need of assembling these prompts at runtime: https://github.com/orgs/langfuse/discussions/1660

**Reply by ajram23** [reply]
@marcklingen @hassiebp more progress. Most of it works. As @hassiebp mentioned choose the outer most prompt and that did the trick. However there is another issue not all prompts are being linked. I am missing a few using PromptTemplate.from_template. Do you folks want me to fire off a trace with Debug=True enabled so you can check it at your end? Just 3 more to go for me.

**Reply by hassiebp** [reply]
That would be awesome @ajram23 - and thanks for your support here. Helps a lot to get as many cases covered as possible.

**Reply by ajram23** [reply]
@hassiebp Done. Sent you an email with the link to the trace and the corresponding python code. Hope it helps!

**Reply by marcklingen** [reply]
thanks @ajram23 for the additional context and the thread tracking interoperability of compile and get_langchain_prompt methods: https://github.com/orgs/langfuse/discussions/3401

**Reply by ajram23** [reply]
@marcklingen @hassiebp thanks for listening as always! I know you folks will get to the bottom of this!

**Reply by hassiebp** [reply]
@ajram23 I have just [released a patch](https://github.com/langfuse/langfuse-python/releases/tag/v2.50.1) that should fix prompt linking for your chain where generations are occurring one level deeper in the tracing tree than the prompt template rendering. Could you please upgrade and try again? 🙏🏾

**Reply by ajram23** [reply]
@hassiebp no such luck. See trace ID 093ba2c5-f446-4768-b53e-cb0fe1629aeb

**Reply by hassiebp** [reply]
🤔 I've created a chain that is similar to the example you have emailed me. When using `with_config`, one has to be mindful that the Runnable is wrapped in another Runnable, so the `langfuse_prompt` must be passed under the `metadata` key as well in the `with_config`. 

Here is a full example where the generation is nested, and [here is a link to the resulting trace](https://staging.langfuse.com/project/7a88fb47-b4e2-43b8-a06c-a5ce950dc53a/traces/8a86cba6-25c0-4c1a-b0d4-ad5d93b45133):

```python
from langfuse import Langfuse
from langchain_openai import ChatOpenAI
from langchain_core.pydantic_v1 import BaseModel, Field
from langchain_core.prompts import PromptTemplate

class Joke(BaseModel):
    setup: str = Field(description="The setup of the joke")
    punchline: str = Field(description="The punchline to the joke")

langfuse = Langfuse()
trace_name = "test_complex_chain"

# Create prompts
joke_prompt_name = "joke_prompt_" + create_uuid()[:8]
joke_prompt_string = "Tell me a joke involving the animal {{animal}}"

langfuse.create_prompt(
    name=joke_prompt_name,
    prompt=joke_prompt_string,
    labels=["production"],
)

# Get prompts
langfuse_joke_prompt = langfuse.get_prompt(joke_prompt_name)
langchain_joke_prompt = PromptTemplate.from_template(
    langfuse_joke_prompt.get_langchain_prompt(),
).with_config(
    {
        "run_name": "joke_prompt",
        "metadata": {"langfuse_prompt": langfuse_joke_prompt},
    }
)

# Create chain
llm = (
    ChatOpenAI(model="gpt-3.5-turbo-0125", temperature=0)
    .with_structured_output(Joke)
    .with_config({"run_name": "llm"})
)

chain = langchain_joke_prompt | llm

# Run chain
langfuse_handler = CallbackHandler(debug=True)
chain.invoke(
    {"animal": "dog"},
    config={
        "callbacks": [langfuse_handler],
        "run_name": trace_name,
    },
)

langfuse_handler.flush()
```

**Reply by ajram23** [reply]
@hassiebp We might be going around in circles here (and mistake could well be mine). I suggest jumping on a zoom call if needed. Here are something that don't add up from end. 1. If I am sending you data in the trace, shouldn't you be able to look in your servers to see if you are getting the meta data or not? 2. Even in your example, I don't see a link to a prompt. May be I am mistaken?
<img width="927" alt="Screenshot 2024-09-19 at 10 17 22 AM" src="https://github.com/user-attachments/assets/e94a6920-4d69-4d22-9005-91a363eba735">. In my case when it works I see this 
<img width="538" alt="Screenshot 2024-09-19 at 10 19 55 AM" src="https://github.com/user-attachments/assets/1cc295a3-680d-4770-9301-6ea34582aaa2">

**Reply by hassiebp** [reply]
Yes, let's hop on a call, will email you an invite 👍🏾 

> If I am sending you data in the trace, shouldn't you be able to look in your servers to see if you are getting the meta data or not

The issue here is on the SDK that do not append the prompt information to our server calls. So to debug, we need visibility on the specific chain running.

> Even in your example, I don't see a link to a prompt. May be I am mistaken?

Apologies, the public view cuts links to other features on the platform. For logged in users, here's the view:
![CleanShot 2024-09-19 at 16 27 24](https://github.com/user-attachments/assets/fd637bfa-abca-422e-8b39-6a6181a9d278)

**Reply by ajram23** [reply]
@hassiebp thanks for a great call! Much appreciated and glad we found a solution! Will wait for the helper for compile/langchain and then ship this! @marcklingen thank you as well!
